lib/user_space/user_x86_64_atomic.h

```diff
@@ -130,48 +141,133 @@ static inline int64_t m0_atomic64_add_return(struct m0_atomic64 *a,
 
  @return new value of atomic counter
  */
-static inline int64_t m0_atomic64_sub_return(struct m0_atomic64 *a,
-						  int64_t delta)
+static inline int64_t m0_atomic64_sub_return(atomic64_t *a, int64_t delta)
 {
+#ifdef CONFIG_X86_64
 	return m0_atomic64_add_return(a, -delta);
+#else  //aarch64
+	int64_t result;
+    uint64_t tmp;
+    asm volatile("// atomic64_sub_return \n"    \
+		"       prfm    pstl1strm, %2\n"        \
+		"1:     ldxr    %0, %2\n"               \
+		"       sub     %0, %0, %3\n"           \
+		"       stlxr    %w1, %0, %2\n"         \
+		"       cbnz    %w1, 1b\n"              \
+		"       dmb ish"                        \
+        : "=&r" (result), "=&r" (tmp), "+Q" (a->a_value)        \
+        : "Ir" (delta)                          \
+        : "memory");
+    return result;
+#endif
 }
 
+/**
+ atomically increment counter
+
+ @param a pointer to atomic counter
+
+ @return none
+ */
+static inline void m0_atomic64_inc(struct m0_atomic64 *a)
+{
+#ifdef CONFIG_X86_64
+	asm volatile("lock incq %0"
+				: "=m" (a->a_value)
+				: "m" (a->a_value));
+#else  //aarch64
+	 m0_atomic64_add(a, (int64_t)1);
+#endif
+}
+
+/**
+ atomically decrement counter
+
+ @param a pointer to atomic counter
+
+ @return none
+ */
+static inline void m0_atomic64_dec(struct m0_atomic64 *a)
+{
+#ifdef CONFIG_X86_64
+	asm volatile("lock decq %0"
+		     : "=m" (a->a_value)
+		     : "m" (a->a_value));
+
+#else  //aarch64
+	 m0_atomic64_sub(a, (int64_t)1);
+#endif
+}
+
+
 static inline bool m0_atomic64_inc_and_test(struct m0_atomic64 *a)
 {
+#ifdef CONFIG_X86_64
 	unsigned char result;
-
 	asm volatile("lock incq %0; sete %1"
 		     : "=m" (a->a_value), "=qm" (result)
 		     : "m" (a->a_value) : "memory");
 	return result != 0;
+#else
+	return (m0_atomic64_add_return(a, 1) == 0);
+#endif
 }
 
 static inline bool m0_atomic64_dec_and_test(struct m0_atomic64 *a)
 {
+#ifdef CONFIG_X86_64
 	unsigned char result;
-
 	asm volatile("lock decq %0; sete %1"
 		     : "=m" (a->a_value), "=qm" (result)
 		     : "m" (a->a_value) : "memory");
 	return result != 0;
+#else
+	return (m0_atomic64_sub_return(a, 1) == 0);
+#endif
+
 }
 
 static inline bool m0_atomic64_cas(int64_t * loc, int64_t oldval, int64_t newval)
-{
+{									
+#ifdef CONFIG_X86_64
 	int64_t val;
-
 	M0_CASSERT(8 == sizeof oldval);
-
 	asm volatile("lock cmpxchgq %2,%1"
 		     : "=a" (val), "+m" (*(volatile long *)(loc))
 		     : "r" (newval), "0" (oldval)
 		     : "memory");
 	return val == oldval;
+#else		 //aarch64	
+	unsigned long tmp, old=0;		
+
+	asm volatile(						\
+	"	prfm	pstl1strm, %[v]\n"			\
+	"1:	ldxr\t%[old], %[v]\n"			\
+	"	eor	%[tmp], %[oldval], %[old]\n"		\
+	"	cbnz	%[tmp], 2f\n"				\
+	"	stxr\t%w[tmp], %[newval], %[v]\n"		\
```

Comment:
can the "\t" be replaced with real "TAB" or "SPACE"?

Reply:
removed /t, replaced with tab